[rpc] various fixes for ProcessGroupAgent

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34943 [rpc] various fixes for ProcessGroupAgent**

Various small changes to address @jjlilley and @mrshenli's comments on
https://github.com/pytorch/pytorch/pull/34413:
1) Continue trying even if one `agent->send()` fails when cleaning up dist
autograd ctx
2) Use RAII for lock in process group agent `handleSend`
3) Return bool instead of int in `ProcessGroupAgent::handleRecv` to determine
if the count should be incremented
4) Move recvCounts increment in timed out future processing to be within the
block that ensures the future already doesn't have an error.

Differential Revision: [D20506065](https://our.internmc.facebook.com/intern/diff/D20506065/)

rohan-varma committed:
[rpc] various fixes for ProcessGroupAgent

Follow up to address Jeremy's and Shen's comments on
https://github.com/pytorch/pytorch/pull/34413:
1) Continue trying even if one `agent->send()` fails when cleaning up dist
autograd ctx
2) Use RAII for lock in process group agent `handleSend`
3) Return bool instead of int in `ProcessGroupAgent::handleRecv` to determine
if the count should be incremented
4) Move recvCounts increment in timed out future processing to be within the
block that ensures the future already doesn't have an error.

Differential Revision: [D20506065](https://our.internmc.facebook.com/intern/diff/D20506065/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] various fixes for ProcessGroupAgent"


Various small changes to address @jjlilley and @mrshenli's comments on
https://github.com/pytorch/pytorch/pull/34413:
1) Continue trying even if one `agent->send()` fails when cleaning up dist
autograd ctx
2) Use RAII for lock in process group agent `handleSend`
3) Return bool instead of int in `ProcessGroupAgent::handleRecv` to determine
if the count should be incremented
4) Move recvCounts increment in timed out future processing to be within the
block that ensures the future already doesn't have an error.

Differential Revision: [D20506065](https://our.internmc.facebook.com/intern/diff/D20506065/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -771,12 +773,12 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
         --clientActiveCalls_;
```
@jjlilley @osalpekar Would any of you know when this condition potentially does not trigger? It seems like this should be the first time we are processing the future, as even if a receive and timeout arrive at the same time, we grab the lock to process both, so only one should process successfully

jjlilley replied:
Can we get in the situation where the future is satisfied by regular IO, when the lock is dropped, racing with this logic?
If anything, we should probably check !complete(). Or call setErrorIfNeeded() rather than setError().

(aside: ideally we should modify future<T> api to return a bool from setErrorIfNeeded(), so that we can make adjust logic if it was needed)

jjlilley commented on torch/csrc/distributed/autograd/context/container.cpp:
```diff
@@ -157,24 +157,38 @@ void DistAutogradContainer::releaseContext(int64_t context_id) {
   // ungraceful shutdown, where we are shutting down RPC and also processing
   // this message in a separate thread concurrently. In this case, don't throw
   // here.
+  std::shared_ptr<rpc::RpcAgent> agent = nullptr;
```
nit: `= nullptr` is implicit

rohan-varma replied:
Removed in the most recent update.

jjlilley reviewed (approved):
LGTM

mrshenli commented on torch/csrc/distributed/autograd/context/container.cpp:
```diff
@@ -157,24 +157,38 @@ void DistAutogradContainer::releaseContext(int64_t context_id) {
+
+  for (const auto& worker_id : workerIds) {
+    try {
+      // Try to send RPC to clear context, though continue attempting to send to
```
though -> through? (not sure)

rohan-varma replied:
I meant for it to be "though", to explain that we want to continue sending in case one fails. I think I can word this better though.

rohan-varma committed:
Update on "[rpc] various fixes for ProcessGroupAgent"


Various small changes to address @jjlilley and @mrshenli's comments on
https://github.com/pytorch/pytorch/pull/34413:
1) Continue trying even if one `agent->send()` fails when cleaning up dist
autograd ctx
2) Use RAII for lock in process group agent `handleSend`
3) Return bool instead of int in `ProcessGroupAgent::handleRecv` to determine
if the count should be incremented
4) Move recvCounts increment in timed out future processing to be within the
block that ensures the future already doesn't have an error.

Differential Revision: [D20506065](https://our.internmc.facebook.com/intern/diff/D20506065/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] various fixes for ProcessGroupAgent"


Various small changes to address @jjlilley and @mrshenli's comments on
https://github.com/pytorch/pytorch/pull/34413:
1) Continue trying even if one `agent->send()` fails when cleaning up dist
autograd ctx
2) Use RAII for lock in process group agent `handleSend`
3) Return bool instead of int in `ProcessGroupAgent::handleRecv` to determine
if the count should be incremented
4) Move recvCounts increment in timed out future processing to be within the
block that ensures the future already doesn't have an error.

Differential Revision: [D20506065](https://our.internmc.facebook.com/intern/diff/D20506065/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] various fixes for ProcessGroupAgent"


Various small changes to address @jjlilley and @mrshenli's comments on
https://github.com/pytorch/pytorch/pull/34413:
1) Continue trying even if one `agent->send()` fails when cleaning up dist
autograd ctx
2) Use RAII for lock in process group agent `handleSend`
3) Return bool instead of int in `ProcessGroupAgent::handleRecv` to determine
if the count should be incremented
4) Move recvCounts increment in timed out future processing to be within the
block that ensures the future already doesn't have an error.

Differential Revision: [D20506065](https://our.internmc.facebook.com/intern/diff/D20506065/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] various fixes for ProcessGroupAgent"


Various small changes to address @jjlilley and @mrshenli's comments on
https://github.com/pytorch/pytorch/pull/34413:
1) Continue trying even if one `agent->send()` fails when cleaning up dist
autograd ctx
2) Use RAII for lock in process group agent `handleSend`
3) Return bool instead of int in `ProcessGroupAgent::handleRecv` to determine
if the count should be incremented
4) Move recvCounts increment in timed out future processing to be within the
block that ensures the future already doesn't have an error.

Differential Revision: [D20506065](https://our.internmc.facebook.com/intern/diff/D20506065/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] various fixes for ProcessGroupAgent"


Various small changes to address @jjlilley and @mrshenli's comments on
https://github.com/pytorch/pytorch/pull/34413:
1) Continue trying even if one `agent->send()` fails when cleaning up dist
autograd ctx
2) Use RAII for lock in process group agent `handleSend`
3) Return bool instead of int in `ProcessGroupAgent::handleRecv` to determine
if the count should be incremented
4) Move recvCounts increment in timed out future processing to be within the
block that ensures the future already doesn't have an error.

Differential Revision: [D20506065](https://our.internmc.facebook.com/intern/diff/D20506065/)

[ghstack-poisoned]
